ksql-engine/src/main/java/io/confluent/ksql/function/UdfLoader.java

```diff
@@ -0,0 +1,154 @@
+        Files.find(pluginDir.toPath(), 1, (path, attributes) -> path.toString().endsWith(".jar"))
+            .map(path -> UdfClassLoader.newClassLoader(path, parentClassLoader, blacklist))
+            .forEach(this::loadUdfs);
+      } catch (IOException e) {
+        logger.error("Failed to load UDFs from location {}", pluginDir, e);
+      }
+    }
+  }
+
+
+  private void loadUdfs(final ClassLoader loader) {
+    new FastClasspathScanner()
+        .overrideClassLoaders(loader)
+        .ignoreParentClassLoaders()
+        // if we are loading from the parent classloader then restrict the name space to only
+        // jars/dirs containing "ksql-engine". This is so we don't end up scanning every jar
+        .filterClasspathElements(
+            name -> parentClassLoader != loader || name.contains("ksql-engine"))
+        .matchClassesWithMethodAnnotation(Udf.class,
+            (theClass, executable) -> {
+              final UdfDescription annotation = theClass.getAnnotation(UdfDescription.class);
+              if (annotation != null) {
+                final Method method = (Method) executable;
+                try {
+                  final UdfInvoker udf = compiler.compile(method, loader);
+                  addFunction(annotation, method, udf);
+                } catch (final KsqlException e) {
+                  if (parentClassLoader == loader) {
+                    throw e;
+                  } else {
+                    logger.warn("Failed to add UDF to the MetaStore. name={} method={}",
+                        annotation.name(),
+                        method,
+                        e);
+                  }
+                }
+              }
+            })
+        .scan();
+  }
+
+  private void addFunction(final UdfDescription annotation,
+                           final Method method,
+                           final UdfInvoker udf) {
+    metaStore.addFunction(new KsqlFunction(
+        SchemaUtil.getSchemaFromType(method.getReturnType()),
+        Arrays.stream(method.getGenericParameterTypes())
+            .map(SchemaUtil::getSchemaFromType).collect(Collectors.toList()),
+        annotation.name(),
+        PluggableUdf.class,
+        () -> {
+          try {
+            return new PluggableUdf(udf, method.getDeclaringClass().newInstance());
+          } catch (Exception e) {
+            throw new KsqlException("Failed to create instance for UDF="
+                + annotation.name()
+                + ", method=" + method,
+                e);
+          }
+        }));
+  }
+
+  public static UdfLoader newInstance(final KsqlConfig config,
+                                      final MetaStore metaStore,
+                                      final String ksqlInstallDir
+  ) {
+    final Boolean loadCustomerUdfs = config.getBoolean(KsqlConfig.KSQL_ENABLE_UDFS);
+    final File pluginDir = new File(ksqlInstallDir, "ext");
+
+    Preconditions.checkArgument(!loadCustomerUdfs || pluginDir.isDirectory(),
```

Comment:
Do we really want to fail if someone enables UDFs but the /ext dir doesn't exist? I'd say log a warning and carry on. No directory should be the same as an empty directory.

Reply:
yeah - fair enough